File: go/test/alpha_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/rest-countries-sdk/go"
	"github.com/voxgig-sdk/rest-countries-sdk/go/core"

	vs "github.com/voxgig-sdk/rest-countries-sdk/go/utility/struct"
)

func TestAlphaEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Alpha(nil)
		if ent == nil {
			t.Fatal("expected non-nil AlphaEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := alphaBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "alpha." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set RESTCOUNTRIES_TEST_ALPHA_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		alphaRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.alpha", setup.data)))
		var alphaRef01Data map[string]any
		if len(alphaRef01DataRaw) > 0 {
			alphaRef01Data = core.ToMapAny(alphaRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = alphaRef01Data

		// LOAD
		alphaRef01Ent := client.Alpha(nil)
		alphaRef01MatchDt0 := map[string]any{}
		alphaRef01DataDt0Loaded, err := alphaRef01Ent.Load(alphaRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if alphaRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func alphaBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "alpha", "AlphaTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read alpha test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse alpha test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"alpha01", "alpha02", "alpha03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("RESTCOUNTRIES_TEST_ALPHA_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"RESTCOUNTRIES_TEST_ALPHA_ENTID": idmap,
		"RESTCOUNTRIES_TEST_LIVE":      "FALSE",
		"RESTCOUNTRIES_TEST_EXPLAIN":   "FALSE",
		"RESTCOUNTRIES_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["RESTCOUNTRIES_TEST_ALPHA_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["RESTCOUNTRIES_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["RESTCOUNTRIES_APIKEY"],
			},
			extra,
		})
		client = sdk.NewRestCountriesSDK(core.ToMapAny(mergedOpts))
	}

	live := env["RESTCOUNTRIES_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["RESTCOUNTRIES_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/AlphaEntityTest.php
<?php
declare(strict_types=1);

// Alpha entity test

require_once __DIR__ . '/../restcountries_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class AlphaEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = RestCountriesSDK::test(null, null);
        $ent = $testsdk->Alpha(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = alpha_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "alpha." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set REST_COUNTRIES_TEST_ALPHA_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $alpha_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.alpha")));
        $alpha_ref01_data = null;
        if (count($alpha_ref01_data_raw) > 0) {
            $alpha_ref01_data = Helpers::to_map($alpha_ref01_data_raw[0][1]);
        }

        // LOAD
        $alpha_ref01_ent = $client->Alpha(null);
        $alpha_ref01_match_dt0 = [
            "id" => $alpha_ref01_data["id"],
        ];
        $alpha_ref01_data_dt0_loaded = $alpha_ref01_ent->load($alpha_ref01_match_dt0, null);
        $alpha_ref01_data_dt0_load_result = Helpers::to_map(is_object($alpha_ref01_data_dt0_loaded) && method_exists($alpha_ref01_data_dt0_loaded, 'data_get') ? $alpha_ref01_data_dt0_loaded->data_get() : $alpha_ref01_data_dt0_loaded);
        $this->assertNotNull($alpha_ref01_data_dt0_load_result);
        $this->assertEquals($alpha_ref01_data_dt0_load_result["id"], $alpha_ref01_data["id"]);

    }
}

function alpha_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/alpha/AlphaTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = RestCountriesSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["alpha01", "alpha02", "alpha03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("REST_COUNTRIES_TEST_ALPHA_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "REST_COUNTRIES_TEST_ALPHA_ENTID" => $idmap,
        "REST_COUNTRIES_TEST_LIVE" => "FALSE",
        "REST_COUNTRIES_TEST_EXPLAIN" => "FALSE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["REST_COUNTRIES_TEST_ALPHA_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["REST_COUNTRIES_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
            ],
            $extra ?? [],
        ]);
        $client = new RestCountriesSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["REST_COUNTRIES_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["REST_COUNTRIES_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
